[ 1단계 - 행운의 로또 미션 ] 오거스(오명석) 미션 체줄합니다.

ohgus opened the pull request:
## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.
- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트  
<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->
- [X] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [X] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [X] 테스트 코드는 모두 정상적으로 실행되나요?
- [ ] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: 

## 리뷰 요청 & 논의하고 싶은 내용  

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
- depth1 만들기
- 매개변수 2개 이하로 받기
- class가 꼭 필요한가?

안녕하세요 시지프! 오거스 입니다 😃
우선 저희 페어는 이번 미션에서 최대한 클래스를 사용하지 않는 방식으로 구현하고자 했습니다. 처음에는 몇개의 클래스를 만들기도 했지만 리팩토링하는 과정에서 클래스가 역할이 없다고 판단해 클래스를 전부 제거하고 최대한 함수만을 이용해 미션을 구현하게 되었습니다.
저희가 이번 미션에서 가장 어렵게 느꼈던 부분은 프로그래밍 요구사항중 depth1과 매개변수 2개 이하로 받기 였습니다.
depth1을 충족하기 위해서 코드를 분리하면 필요한 매개변수가 2개 이상으로 많아지는 경우가 생겼기 때문입니다... 🥲
계속 여러 방식을 시도하다가 결국에는 두 가지 조건을 만족할 수 있게 리팩토링할 수 있었지만 정말 힘들었어요 ㅜ
이 과정에서 배운 부분은 무작정 분리하는 것보다 어떻게 하면 적절하게 분리할 수 있을지를 고민해야 한다고 느낀것 같아요.
최대한 단위를 작게 쪼개어 보면서 어느 부분까지 하나의 흐름으로 생각하고 분리할 것인지 그 기준을 찾는 것이 정말 힘들었던것 같아요!

### 2) 이번 리뷰를 통해 논의하고 싶은 부분 
- TDD
  - TDD를 지키는 것이 정말 어려웠어요...
  - TDD를 왜 해야 하는지 솔직히 잘 모르겠어요 (기능 -> 테스트 순으로 해도 충분하지 않을까? 라고 생각이 들어요)
  - 익숙하지 않아서 그런지 몰라도 TDD를 하는데 너무 오래 걸려서 나중에는 제출을 위해 기능 구현을 우선시 했어요
  - TDD가 개발자에게 꼭 필요한 역량일까요?
- 저희가 구현한 코드가 관심사에 맞게 잘 묶여 있는지 궁금해요
  - 관련된 함수끼리 같은 파일에 잘 모여져 있는지 궁금해요
- 네이밍이 적절한지 궁금해요
- 다른 개발자가 보았을때 가독성 면에서 어느정도 수준일지 궁금해요
- constant파일을 만들 때 서로 관련이 없지만 별도의 파일로 분리하기 애매한 경우 어떻게 하는 것이 좋을지 모르겠어요
  - 파일 안에 상수 두 개만 들어간다고 해도 별도의 파일로 분리하는 것이 좋을까요? 아니면 비슷한 상황에 있는 상수들을 모아 하나의 파일로 관리하는 것이 좋을까요?


---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계
- [x] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [ ] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [x] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [x] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계 
- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가? 
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

yeji0214 committed:
docs: Update README.md Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 구입 금액 입력 받기

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 당첨 번호 입력 받기

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 보너스 번호 입력 받기

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 재시작 여부 입력 받기

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 구입 금액 예외 처리

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 랜덤 숫자 생성 구현

Co-authored-by: ohgus <ohgus@users.noreply.github.com>

yeji0214 committed:
feat: 1~45 사이의 6개의 숫자로 이루어진 로또 번호 발생

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 수량만큼 로또 번호 발행 및 출력

Co-authored-by: ohgus <ohgus@users.noreply.github.com>

yeji0214 committed:
feat: 당첨 번호 유효성 검사

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 보너스 번호 유효성 검사

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 당첨 번호와 보너스 번호 한 번에 처리

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 당첨 기준에 따라 당첨 개수 세기

Co-authored-by: ohgus <ohgus@users.noreply.github.com>

yeji0214 committed:
feat: 당첨 내역 출력

yeji0214 committed:
feat: 총 상금 계산

Co-authored-by: ohgus <ohgus@users.noreply.github.com>

yeji0214 committed:
feat: 수익률 계산

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 재시작/종료 여부 입력 받기

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 예외 발생 시, 에러 메시지 출력 후 그 부분부터 다시 입력 받기

Co-authored-by: ohgus <ohgus@users.noreply.github.com>

yeji0214 committed:
refactor: totalPrize 로직 수정 및 테스트 추가

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
refactor: 구조 수정

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
style: import문 정렬

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
refactor: export문 통일화

yeji0214 committed:
refactor: 함수 depth 1로 수정

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
style: 입출력 메시지 줄바꿈

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
refator: 재시작 입력 수정

Co-authored-by: ohgus <anakin3740@gmail.com>

yeji0214 committed:
feat: 에러 메시지 수정

Co-authored-by: ohgus <anakin3740@gmail.com>

woowa-euijinkk commented on src/step1-index.js:
```diff
@@ -1,4 +1,35 @@
+    return answer;
+  }
+
+  endGame(answer) {
```
.

ohgus replied:
혹시 이건 어떤 리뷰인지 여쭤봐도 될까요...?

woowa-euijinkk replied:
앗 제가 체크해둔다는걸, 수정을 안했군요.

endGame 인데 조건부로 게임이 끝나는게 이상합니다! 
함수명 or 내부 기능 수정이 필요합니다

woowa-euijinkk commented on src/LottoMachine.js:
```diff
@@ -0,0 +1,16 @@
+    lotto.add(randomNumber);
+  }
+
+  return Array.from(lotto).sort((a,b) => a - b);
```
generateLotto 인데 sort 역할까지 하고 있습니다. 역할 분리가 더 필요해보입니다!

ohgus replied:
정말 간단한 동작의 경우도 (예를 들어 sort 같은 역할이요!) 역시 따로 분리하는 것이 좋을까요? 
뭔가 너무 과한 분리가 아닐까 고민했습니다!

woowa-euijinkk replied:
아하, 좋은 고민이네요. 그 기준점을 잘 고민해보시면 좋겠네요!

누군가에겐 당연한 기능이고, 누군가에겐 당연하지 않은 기능이라 상황에 따라 다를 것 같아요. 

- 오름차순이 a-b 인지 b-a 인지 헷갈린다면, 기능 분리 되면 더 좋을거같아요.
- `Array.from(lotto).sort((a,b) => a - b)` 를 다 읽는 것보다 sortByAsc(lotto) 를 읽는게 더 쉽긴 하겠어요.

저도 무조건 해야 한다고 생각하진 않고, 오거스의 의견대로 진행하시면 되겠습니다!

woowa-euijinkk commented on src/Calculator.js:
```diff
@@ -0,0 +1,50 @@
+import { WINNING, KEYS } from "./constant/lotto.js";
+
+const Calculator = {
+  winningCount(lottos, { winning, bonus }) {
```
함수명이 이상합니다.

함수는 동사이고, 변수는 명사여야 합니다!
전체적으로 함수명, 변수명을 더 잘 짓기 위한 노력이 필요해보여요! 
보기에 어색해서, 가독성이 좋지 않은 이름들이 많습니다.

ohgus replied:
네이밍을 전체적으로 수정하겠습니다!

woowa-euijinkk commented on src/Calculator.js:
```diff
@@ -0,0 +1,50 @@
+    const winningCount = { [KEYS.FIRST]: 0, [KEYS.SECOND]: 0, [KEYS.THIRD]: 0, [KEYS.FOURTH]: 0, [KEYS.FIFTH]: 0 };
+
+    lottos.forEach((lotto) => {
+      const matchCount = lotto.filter((num) => winning.includes(num)).length;
```
변수명이 동사형이라 어색합니다.
matchedCount 가 어떨까요?

woowa-euijinkk commented on src/constant/lotto.js:
```diff
@@ -0,0 +1,42 @@
+    [KEYS.FIRST]: {
+        MATCH: 6,
+        LABEL: '6개 일치',
+        PRIZES: 2_000_000_000
```
👍

woowa-euijinkk commented on src/constant/lotto.js:
```diff
@@ -0,0 +1,42 @@
+export const WINNING = {
+    [KEYS.FIRST]: {
+        MATCH: 6,
+        LABEL: '6개 일치',
```
LOTTO.LENGTH 가 변수인데 반해, 여기는 변수화가 되어있지 않네요!

ohgus replied:
```js
const MATCHES = {
  [KEYS.FIRST]: 6,
  [KEYS.SECOND]: 5,
  [KEYS.THIRD]: 5,
  [KEYS.FOURTH]: 4,
  [KEYS.FIFTH]: 3,
};

export const WINNING = {
  [KEYS.FIRST]: {
    MATCH: MATCHES[KEYS.FIRST],
    LABEL: `${MATCHES[KEYS.FIRST]}개 일치`,
    PRIZES: 2_000_000_000,
  },
```

이런식의 변수화를 말씀하시는 걸까요?

woowa-euijinkk replied:
맞아요!

woowa-euijinkk commented on src/Calculator.js:
```diff
@@ -0,0 +1,50 @@
+  },
+
+  increaseCount(winningCount, matchKey) {
+    if (matchKey) {
```
`increaseCount` 함수는 count 를 증가시키는 역할만 하면 될 것 같은데요.
조건을 판단해서 증가시키고 있군요. 

정확한 이름은 `increaseCountOrNot` 정도 일 것 같아요. 그런데 조건문이 여기 함수에서 책임을 가지는게 좋을까요, 밖에서 판단하는게 좋을까요?

ohgus replied:
원래는 저 조건문을 밖에서 사용하려 했지만 depth1을 지키기 위해서 `increaseCount` 안으로 넣게 되었어요... `getMatchKey` 함수가 꽝에 대한 값은 반환해주지 않다 보니 이런 일이 생긴것 같은데 혹시 꽝에 대한 처리도 필요할까요?

ohgus replied:
만약 하나의 로또 결과를 체크하는 로직을 분리한다면 `increaseCount` 밖으로 분리할 수 있을것 같아요!

woowa-euijinkk replied:
> 혹시 꽝에 대한 처리도 필요할까요?

있으면 사용이 편할거같아요.

---

오오 뎁스1을 지키기 위해서 많이 노력해주셨군요. 이번 미션은 하드하게 프로그래밍 요구사항이 설정되어 있는데요. 덕분에 함수 분리 많이 하면서 고민 많이 하게 되신 것 같아요. 

추가로 depth 1을 지키기 위해서 early return 하는 방법도 알려드립니다.
이번 같이 간단한 경우엔 early return 하지 않는게, 적절할 수도 있겠지만, 다음번에 자주 썼으면 하여 공유드려요.

```js
if (!shouldIncrease()) {
  return;
}
increaseCount()
```

woowa-euijinkk commented on src/Calculator.js:
```diff
@@ -0,0 +1,50 @@
+    const winningCount = { [KEYS.FIRST]: 0, [KEYS.SECOND]: 0, [KEYS.THIRD]: 0, [KEYS.FOURTH]: 0, [KEYS.FIFTH]: 0 };
+
+    lottos.forEach((lotto) => {
+      const matchCount = lotto.filter((num) => winning.includes(num)).length;
+      const hasBonus = this.hasBonus(lotto, bonus);
```
아래는 함수화 되어 있고, 위는 함수화 되어 있지 않습니다.
이를 추상화 계층이 맞지 않다고 표현합니다. 
추상화 계층이 맞지 않으면, 코드를 읽기 어색합니다!

추상화 할거면, 둘다 추상화 하고, 하지 않을거면 둘다 하지 않는 것이 좋아보여요

ohgus replied:
아하 확실히 일관성이 떨어지는 것 같아요! 아래 코드들은 요구사항을 맞추기 위해서 수정을 거치다 보니 함수화가 되었는데 위 코드는 생각하지 못한 것 같아요!

woowa-euijinkk commented on src/Calculator.js:
```diff
@@ -0,0 +1,50 @@
+    const winningCount = { [KEYS.FIRST]: 0, [KEYS.SECOND]: 0, [KEYS.THIRD]: 0, [KEYS.FOURTH]: 0, [KEYS.FIFTH]: 0 };
+
+    lottos.forEach((lotto) => {
+      const matchCount = lotto.filter((num) => winning.includes(num)).length;
+      const hasBonus = this.hasBonus(lotto, bonus);
+
+      const matchKey = this.getMatchKey(matchCount, hasBonus);
+      this.increaseCount(winningCount, matchKey);
```
`winningCount` 의 역할이 많아 보입니다.
하나의 로또에 대해서 계산하는 함수는 별도로 분리하면 어떨까요?

ohgus replied:
열심히 분리 해보겠습니다!

woowa-euijinkk commented on src/Calculator.js:
```diff
@@ -0,0 +1,50 @@
+    return;
+  },
+
+  totalPrize(winningCount) {
```
마찬가지로 변수명이 명사형이라 어색합니다!

woowa-euijinkk commented on src/Calculator.js:
```diff
@@ -0,0 +1,50 @@
+  },
+
+  yieldRate(amount, totalPrize) {
+    return ((totalPrize / amount) * 100).toFixed(1);
```
소수점 한자리까지 잘라서 보여주는건 UI 인가요? 도메인인가요?
이 함수의 책임이 맞나요?
cf) A 페이지에서는 소수점 두자리까지, B페이지에서는 0번째까지 보여줘야하는 경우

ohgus replied:
출력 로직에서 소수점 처리를 하는 것이 맞는 것 같습니다!

woowa-euijinkk commented on src/step1-index.js:
```diff
@@ -1,4 +1,35 @@
+  }
+
+  async start() {
+    while(this.#running) {
```
start 이지만, "무한 start" 을 포함한다는 측면에서 이름과 역할이 적절하지 않습니다.

woowa-euijinkk commented on src/step1-index.js:
```diff
@@ -1,4 +1,35 @@
+    const answer = await InputHandler.getRestartAnswer();
+
+    return answer;
+  }
```
이건 왜 함수화 하셨는지 의도가 궁금합니다!

woowa-euijinkk commented on __tests__/Calculator.test.js:
```diff
@@ -0,0 +1,32 @@
+    const winningCount = { [KEYS.FIRST]: 0, [KEYS.SECOND]: 0, [KEYS.THIRD]: 2, [KEYS.FOURTH]: 0, [KEYS.FIFTH]: 0 };
+    const totalPrize = Calculator.totalPrize(winningCount);
+
+    expect(totalPrize).toBe(3_000_000);
```
변수화가 필요하겠군요!

ohgus replied:
toBe 값을 말씀하시는 걸까요??

woowa-euijinkk replied:
음.. 깃허브 에러인가, range 가 이상하게 설정되었네요.

3_000_000 이 변수화가 필요하다는 것이었습니다. 
winningCount["2등"] * PRICE{"2등"] 식으로요!

상금 1_500_000 이 변경되면 바로 테스트가 깨지게 되니까요

ohgus replied:
수정했습니다.

[상수 적용](https://github.com/woowacourse/javascript-lotto/pull/363/commits/06619a7ad9318226241123e992187415691b8011)

woowa-euijinkk commented on __tests__/Calculator.test.js:
```diff
@@ -0,0 +1,32 @@
+
+describe('', () => {
+  test.each([
+    [[[1,2,3,4,5,6]], 1, '1등'],
+    [[[1,2,3,4,5,9]], 1, '2등'],
+    [[[1,2,3,4,5,10]], 1, '3등'],
+    [[[1,2,3,4,7,10]], 1, '4등'],
+    [[[1,2,3,8,7,10]], 1, '5등'],
```
꽝에 대한 테스트 케이스는 없는 것 같아요!

ohgus replied:
winningCount에 꽝에 대한 처리가 없기 때문에 케이스를 고려하지 않았는데 역시 추가하는게 좋을까요?

woowa-euijinkk replied:
아하, 
그러게요. 현재 요구사항에는 없기 때문에, 무조건적으로 필요한 건 아니겠네요.
데이터 모델상 있는 것이 자연스러워 보이긴 해서 말씀드리긴 했어요.
넘어가셔도 될거같습니다~!

woowa-euijinkk commented on __tests__/Calculator.test.js:
```diff
@@ -0,0 +1,32 @@
+  })
+
+  test('수익률을 계산한다.', () => {
+    const amount = '5000';
```
amount 는 왜 문자열로 표현하셨을까요?~

ohgus replied:
실제 로직에서 문자열을 넘버 타입으로 변환하지 않고 사용했기 떄문에 똑같이 적용하는 것이 맞지 않을까라고 생각해서 문자열을 입력했습니다.

구현중에 고민했던 부분이긴 한데 문자열을 이용해서 연산을할 때 자동으로 형변환을 해서 연산이 이루어지는데 연산전에 넘버 타입으로 변환이 꼭 필요한지 궁금합니다!

woowa-euijinkk replied:
아하 고민이 되겠어요.

일단 getYieldRate 에서는 숫자 와 숫자를 연산시키는 로직이기 때문에, 문자열로 들어가는 것이 어색합니다.
`getYieldRate` 가 요구하는 타입에 맞게 넣어주는게 단위 테스트 측면에서 적절할 것 같습니다. 

더하여, 실제 로직에서 변환하지 않고 넣어주는 것이 타입 안정성을 해칩니다.

woowa-euijinkk commented on __tests__/generateLotto.test.js:
```diff
@@ -0,0 +1,9 @@
+import generateLotto from '../src/LottoMachine.js';
+
+
+test(`1~45 사이의 ${LOTTO.LENGTH}개의 숫자로 이루어진 로또를 생성한다.`, () => {
+  const lotto = generateLotto();
+
+  expect(lotto.length).toBe(LOTTO.LENGTH);
+})
```
- 중복 없다는 요구사항에 대한 테스트가 부족합니다!
- 45 변수화가 필요하겠네요!

ohgus replied:
입력이 아닌 값에 대한 중복 테스트는 필요없지 않을까 생각했는데 하는게 맞을까요?

woowa-euijinkk replied:
엇 오거스, 어떤 의미인지 잘 이해를 못했어요.
더 자세히 설명해주실 수 있나요?

ohgus replied:
앗 저는 중복을 검증한다는 의미가 사용자가 입력하는 로또 번호에만 해당한다고 생각했어요!
저희가 생성하는 로직은 당연히 중복된 번호가 없다고 생각했기 때문입니다!

그런데 시지프의 리뷰를 보고 생각이 조금 바뀌었어요... 
테스트라는 것이 제가 구현한 코드가 예상한 대로 제대로 동작하고 있는지를 검증하기 위함이니까 필요할 것 같습니다!

woowa-euijinkk commented on src/util/random.js:
```diff
@@ -0,0 +1,11 @@
+import { LOTTO } from "../constant/lotto.js";
+
+const getRandomNumber = () => {
+  return (
+    Math.floor(
+      Math.random() * (LOTTO.MAX_RANDOM_VALUE - LOTTO.MIN_RANDOM_VALUE + 1)
+    ) + LOTTO.MIN_RANDOM_VALUE
+  );
+};
```
이름이 getRandomNumber 라는 general 한 이름이지만, 1~45 까지의 랜덤 정수만 뽑는다는 측면에서 이상한 것 같습니다.
함수를 다시 설계해보면 좋을 것 같습니다!

아예 getLottoNumber 함수로 사용하고, util이 아닌 domain 에 포함시켜버리거나, 아예 재사용 가능하게 만들어서 util 에 유지하거나, 두가지 방법이 있을 것 같아요.

ohgus replied:
만약 안에서 사용되는 상수의 이름이 `LOTTO`와 관련이 없고 그냥 단순히 `MIN_RANDOM_VALUE`, `MAX_RANDOM_VALUE`로 바뀌어도 재사용 가능성이 없어질까요?

woowa-euijinkk replied:
네 저는 그렇게 생각하는데, 오거스는 왜 재사용이 있다고 생각하시나요?

ohgus replied:
랜던 넘버함수가 지금은 로또 번호를 생성하고 있지만 범위를 바꿔서 다른 용도로도 사용될 수 있지 않을까 생각했습니다!
상수의 이름에 LOTTO가 붙어 있어서 지금은 로또에만 사용되는 것이 맞는 것 같지만요!

만약 함수의 이름이 `getRandomNumberInRange`라는 이름으로 바뀌고, 상수의 이름도 좀더 범용적인 이름으로 바뀐다면 재사용이 가능한 함수가 될까요?

woowa-euijinkk replied:
lotto 이름이 붙는다면, 로또 내에서만 재사용 가능한 함수가 되는거겠지만요, 오늘 수업하였듯, 45 -> 69로 바뀐다면 재사용이 되지 않습니다. 
1가지 형태로밖에 사용되 수 없으므로 사실상 재사용 가능하다고 부를 수 없겠어요.

`getRandomNumberInRange` 식으로 사용하면 확실히 재사용이 가능하겠죠. 
오늘 수업에 나왔듯, closure 를 통해 high-order function 을 구성하셔도 됩니다.

woowa-euijinkk commented on __tests__/random.test.js:
```diff
@@ -0,0 +1,9 @@
+test(`랜덤으로 생성된 숫자는 ${LOTTO.MIN_RANDOM_VALUE}과 ${LOTTO.MAX_RANDOM_VALUE} 사이여야 한다.`, () => {
+    const randomNumber = getRandomNumber();
+
+    expect(randomNumber).toBeGreaterThanOrEqual(LOTTO.MIN_RANDOM_VALUE);
+    expect(randomNumber).toBeLessThanOrEqual(LOTTO.MAX_RANDOM_VALUE);
```
만약 잘못 계산하여, getRandomNumer 가 0부터 45까지 나오는 함수가 된다면, 이 테스트가 잘못되었음을 알 수 있는 확률은 1/46 이겠네요!

테스트는 결과의 일관성이 보장되어야 한다는 측면에서 적절하지 않을 것 같아요
어떻게 하면 더 정확한 테스트가 가능할까요?

ohgus replied:
확실히 테스트의 결과가 보장되지 않는 느낌인 것 같습니다 ㅜ
더 찾아보고 수정하겠습니다!

woowa-euijinkk replied:
1000개로 테스트해주셨군요. 적절해보입니다!

woowa-euijinkk commented on __tests__/validate.test.js:
```diff
@@ -0,0 +1,71 @@
+    describe('예외 케이스', () => {
+      test(`숫자가 아닌 값인 경우`, () => {
+        const numbers = [1,2,3,4,5,6];
+        const number = '에리얼';
```
변수가 문자열임이 드러나는 이름이어야 하지 않을까요? 
`str` or `notNumber` 등이요!

ohgus replied:
원래는 넘버가 들어와야 하는데 문자열이 들어온 경우라 생각해서 그냥 `number`로 사용했는데 지금 생각해보니 `notNumber` 같은 네이밍이 더 좋았을거 같습니다!

woowa-euijinkk commented on __tests__/validate.test.js:
```diff
@@ -0,0 +1,71 @@
+
+      test(`${LOTTO.MIN_RANDOM_VALUE}~${LOTTO.MAX_RANDOM_VALUE} 사이의 숫자가 아닌 경우`, () => {
+        const numbers = [1,2,3,4,5,6];
+        const number = '46';
```
이렇게 변수화 하는건 어떨까요?

```suggestion
        const number = `${LOTTO.MAX_RANDOM_VALUE + 1}`
```

ohgus replied:
처음에는 예외 케이스 때문에 상수를 추가해야 하나 고민했는데 이런 방법이 있었네요 감사합니다!

woowa-euijinkk commented on __tests__/validate.test.js:
```diff
@@ -0,0 +1,71 @@
+      });
+      
+      test(`중복된 번호가 있는 경우`, () => {
+        const numbers = [1,2,3,4,5,6];
+        const number = '6';
```
이것도 변수화하여 명확히 중복됨을 보여봐요!

ohgus replied:
변수의 이름을 중복된 번호라고 명확히 설명해주는 것을 말씀하시는 걸까요??

woowa-euijinkk replied:
아니요! 
이게 태그가 잘못 걸린건가요? 이상한데 걸린거같네요.

여튼 이런 형태로 중복됨을 보이는 것을 생각했어요

```js
        const duplicatedNumber = 5
        const numbers = [1,2,3,4,duplicatedNumber,duplicatedNumber];
```

ohgus replied:
수정했습니다! 

[중복 변수화](https://github.com/woowacourse/javascript-lotto/pull/363/commits/91adb7b1f964c4c9b26b522ed605d7faed41b6c2)

woowa-euijinkk commented on __tests__/validate.test.js:
```diff
@@ -0,0 +1,71 @@
+import { LOTTO } from "../src/constant/lotto.js";
+import { validateBonusNumber, validateRestart, validateWinningNumbers } from "../src/util/validate.js";
+
+describe('validate', () => {
```
이 테스트가 전체적으로 단위 테스트로 볼 수 있을까요?
validate~ 함수 설계의 문제로 볼 수도 있는데요. 하나의 함수가 여러개의 역할을 맡다보니, 여러개의 기능을 하나의 함수로 테스트 하고 있어요. 

이게 적절하다고 생각하는지, 아닌지 생각해보아요~ (답은 없음!)

woowa-euijinkk commented on src/constant/input.js:
```diff
@@ -0,0 +1,8 @@
+const INPUT_MESSAGE = {
+  AMOUNT: '> 구입금액을 입력해 주세요: ',
+  WINNING_NUMBERS: '\n> 당첨 번호를 입력해 주세요: ',
+  BONUS_NUMBER: '\n> 보너스 번호를 입력해 주세요: ',
```
이 변수 안에 \n 이 있는 것이 적절할까요?

ohgus replied:
좀 애매하다고 생각이 들었어요...
출력할 때 `\n${출력 메시지}`가 맞을지 아니면 `\n`도 하나의 문장으로 보는 것이 맞을지 혹은 하나의 `console.log`를 추가하는 것이 맞을지...

woowa-euijinkk commented on src/constant/lotto.js:
```diff
@@ -0,0 +1,42 @@
+    MIN_RANDOM_VALUE: 1,
+    MAX_RANDOM_VALUE: 45,
+    LENGTH: 6,
+    SPLITTER: ','
```
SPLITTER 가 LOTTO 의 범주에 들어가는 것이 적절한 책임일까욥?~

ohgus replied:
사실 pr에 상수에 관한 질문을 남긴 이유가 이 SPLITTER 떄문이었어요!
아무리 고민해도 어디에 속하는 것이 맞을지 모르겠더라고요...
그래서 로또 번호를 분리할 때 사용하는 것이니 로또에 있어도 괜찮지 않을까? 라고 생각하면서 이곳에 넣었는데 시지프의 리뷰대로 사용하는 곳에서 관리해도 좋을것 같습니다!

woowa-euijinkk replied:
다음에 코멘트 다실때, 각 코멘트에 수정한 commit hash 를 붙여주시면, commit 단위로 하나씩 변경사항 확인하기 편할 것 같아요! 

ex)
[refactor: SPLITTER를 LOTTO에서 분리](https://github.com/woowacourse/javascript-lotto/pull/363/commits/0bfb09f4fead2c3340b66bb545811c57e133c3cf)

ohgus replied:
참고하겠습니다!

woowa-euijinkk reviewed (changes requested):
안녕하세요 오거스!
로또 미션 잘 부탁드려요!~

어려운 프로그래밍 요구사항을 충족하시느라 고생하셨군요. 그 과정에서 배운 점들이 있었다니 다행이에요!

전체적으로 피드백 드리면서, 질문에 답을 드리도록 할게요~~!

## 1. 클래스 대신 함수를 사용하신 점
클래스 사용에 의문을 가지고, 함수를 사용해주신 포인트 좋았어요. 이 기술을 왜 써야할까 의문을 가지는 태도가 항상 필요합니다! 불필요한 기술을 막 도입해선 안되죠. 
그럴 때, 기술 없이 써보기도 하고, 기술을 도입해서 리팩터링 하고 하면서 필요성을 느껴볼 수 있을 것 같아요.
이런 의문을 계속 유지해주시고, 의심하면서 학습하시면 좋겠네요~~

자바스크립트와 클래스의 역사에 대해서 공부해보아요. 
- class 없을 때는 어떻게 코드를 짰을까?
- 은닉을 위해서 이상해보이는(?) 패턴을 쓰는 것을 알게 될텐데, 이건 왜 하는걸까?
- class 를 왜 도입했을까?
- class 는 결국 syntax sugar 라는데 무슨 말일까?
등을 공부해가며 class 의 본질에 대해서 더 생각해보아요.

추가로, 이런 작은 프로그램을 만들땐, 객체와 함수만을 사용해도 충분하다고 생각합니다.
프로그램이 커지면, class 로 상태와 메소드를 한번에 관리해서 모듈화 해야 할 필요성이 생기고, 그 객체끼리 협력해야할 일이 많이 생깁니다. 그럴 때 필요하리라 생각합니다.
ex) Calculator 에서 winningCount 나 상수들을 내부 프로퍼티로 사용해서, 더 응집적으로 협력하게 만들기

## 2. 네이밍 문제
전체적으로 네이밍 문제가 있습니다.
함수는 동사형이어야 하고, 변수는 명사형이어야 합니다.
오거스의 코드는 그렇지 않아서, 읽기가 어색했습니다.

다른 레퍼런스를 많이 보면서, 이런 식으로 이름을 짓는구나! 감을 잡아 보아요. 
https://jacking75.github.io/ETC_good_function_name/
https://brunch.co.kr/@wapj2000/29

## 3. 함수와 변수의 역할
함수, 변수 역할이 명확하지 않는 경우들이 꽤 있습니다.
- 함수명과 함수의 역할이 매칭되지 않음
- 역할이 2개 이상임.

함수의 이름과 역할이 딱 맞아 떨어지게 고민해보면 좋을 것 같아요. 이름이 모호하다는건, 역할이 모호하다는거겠죠.

함수 이름은 협업에 있어서 굉장히 중요합니다.
함수 이름이 적절하지 않아서, 내부 코드 한줄한줄을 봐야 이해할 수 있었어요. 
코드를 빠르게 볼땐, 함수와 변수명만 가지고 읽어나가는데, 그게 어려웠습니다!

## 4. 상수 분리
상수 분리 전체적으로 잘 해주신 것 같아요. 변수화도 신경써주시고, 적절해보입니다.
파일 분리 기준에 대해서 생각해봅시다.

1. 변수는 어디에 위치해야 하는가? -> 유사한 기능 가까이 위치 시키기 (=응집)
2. 유사한 기능이란 무엇을 말하는가? -> LottoStore 에 쓰이는 것이 LottoStore 에 위치시키는게 유사한 것을 모으는 것일까, constant 에 위치시키는 것이 유사한 것일 모으는 것일까. (두가지 관점 모두 가능함. 상황에 따라 다름)
3. 한 곳에서만 사용하는가, 여러 곳에서만 사용하는가?

등을 고민해보고 의사결정 하면 좋겠습니다. 무조건적으로 constant 로 가야한다! 는 절대 아닙니다. 저는 LottoStore 에서만 쓰이는 상수가 있으면 LottoStore 에 함께 넣습니다. 파일 타고 간다는 것은 멀어지는 것이기 때문입니다. 

## 5. TDD
TDD 에 대한 의구심이 많군요! 이번이 처음이라 더 그러신 것 같아요 ㅎㅎ 그리고 처음엔 다 어려운게 당연한 것이라 자연스러운 상태인 것 같아요

1. TDD 를 왜 쓰는지 학습을 해보아요. 단순히 Test First 인가? Test Driven 인가? Test Driven 은 무엇을 의미하나? 단지 테스트를 먼저 쓰기 위함이 아니라, 역할과 책임을 명확히 하고 설계를 잘 하기 위해서 쓰는 방법론 입니다.
2. Test, TDD 등이 처음에 매우 어색할 수 있습니다. 그래서 오히려 인지 비용이 더 발생하고, 시간이 오래 걸리죠. 학습 과정이라서 그렇습니다. 이 과정을 견디지 않으면 나중에 Test 를 해야할 때, Test 비용이 크기 때문에 또 하지 않게 되는 악순환이 반복되죠. 익숙해지면 비용이 줄고, 효용만 남게 되는 상황이 됩니다. 어색하고 어렵기 때문에 왜 써야 하는지 모르겠다는 것은 적절한 이유는 아닌 것 같다고 말씀드리고 싶어요
3. TDD 가 개발자에게 꼭 필요한 역량일까요? -> 꼭 필요하진 않겠죠 ㅎㅎ 하나의 유용한 방법론이니까요. 오거스가 열심히 의심하고, 자신만의 생각을 가져가시면 됩니다. 단, 많은 개발자들이 사랑하는 방법론이니, 긍정적으로 바라볼 필요는 있다는 점!

앞으로도 자주 나오는 키워드일텐데, 왜 쓰는지, 어떤 사고방식인지 이해하는 것이 매우 중요합니다. 완벽하게 하려고 하진 말고, 사고 방식을 이해하는데 초점을 맞춰보아요.

---
## 마치며
다양한 측면에서 피드백을 많이 남겼어요! 
우테코의 시간은 기니까, 천천히 하나씩 학습해가면 좋겠어요. 

수정하다가, 어려운 부분 있으면 더 알려주세요~

ohgus reviewed (commented):
시지프 상세한 리뷰 너무 감사합니다 😃
pr을 올릴 당시에 기억나지 않았던 고민 포인트들 까지 잡아주셔서 어떤 식으로 수정을 해야할지 
조금은? 감이 잡힌 것 같아요!! 열심히 수정해서 다시 리뷰요청 올리겠습니다!! 😁

ohgus committed:
refactor: message에서 newLine 분리

- newLine 상수 생성
- OutputView.js, InputView에 newLine 상수 적용

ohgus committed:
style: prettier 스타일 적용

페어와 프리티어 세팅이 달라서 제 세팅을 적용했습니다.

ohgus committed:
style: 프리티어 적용이 누락된 파일 추가 적용

ohgus commented:
추가로 궁금한 사항이 있습니다!

```
UNDER_MIN_PRICE: `[ERROR] 최소 결제 금액은 ${PRICE.UNIT.toLocaleString()}원 이상입니다.`
```

상수에 `toLocaleString`을 적용해도 될까요?

ohgus committed:
refactor: SPLITTER를 LOTTO에서 분리

ohgus committed:
refactor: validate test 수정

- 상수 적용 가능한 부분 적용
- 변수 이름을 테스트 케이스에 맞게 수정

ohgus committed:
refactor: 테스트 기대 값을 변수화

ohgus committed:
refactor: 상수 적용

- 최소 값, 최대 값

ohgus committed:
refactor: 등수 별 일치 개수를 별도의 상수로 생성

ohgus committed:
refactor: LottoMachine 리팩토링

- getRandomNumber -> getLottoNumber 로 수정 후 LottoMachine 파일로 이동
- 로또 정렬 로직 별도의 함수로 분리

ohgus committed:
refactor: LottoMachine 수정에 따른 리팩토링

- 정렬 로직 테스트 추가
- 생성된 로또 숫자 범위 테스트 강화
- 중복 테스트 추가

ohgus committed:
refactor: 불필요한 함수 제거

ohgus committed:
refactor: Calculator 리팩토링

- 개별 로또에 대한 로직 분리
- increaseCount는 증가만 담당하게 수정
- handleWinningCount 함수 추가
- 명사형으로 되어 있는 함수의 이름을 동사형으로 수정
- matchedCount를 구하는 로직을 함수화

ohgus committed:
refactor: toFixed를 OutputView로 이동

ohgus committed:
refactor: 에러 메시지 수정

- 지폐 단위를 보여주게 수정

ohgus committed:
refactor: 게임 종료 여부를 확인하는 함수 추가

- endGame 함수에서 조건문을 분리

ohgus committed:
test: 꽝에 대한 테스트 추가

ohgus committed:
refactor: winningCount를 복수형으로 변경

ohgus committed:
chore: 누락된 불러오기 추가

ohgus committed:
refactor: OutputView에 모든 로또를 넘겨주고 알아서 출력하게 수정

ohgus committed:
refactor: 관련된 동작을 하나의 함수로 묶어 purchase 함수의 가독성을 향상

ohgus committed:
refactor: Calculator 분리

- 당첨 결과 산정 로직을 LottoCenter 객체에 위임
- 구매한 수량을 계산하는 메서드 추가
- 수량 계산 테스트 추가
- 객체 분리에 따라 테스트도 분리

ohgus committed:
refactor: start 메서드에서 while문을 분리

- run 메서드에서 반복을 담당
- running이 true인 동안 run이 동작하고 start를 계속 호출한다.

ohgus commented:
안녕하세요 시지프!
리뷰하신 내용을 보고 페어와 고민하던 부분들을 많이 찝어주셔서 놀랐습니다 🤔
리뷰를 보고 많이 공부가 되어서 감사드리고 싶어요!! 😁

리뷰하신 내용들을 나름대로 정리하고 반영했습니다!
## 큼직한 변경점
- [x] 네이밍 수정 명사형인 함수의 이름들을 동사형으로 변경했어요.
- [x] 상수를 추가적으로 수정했어요. (파일 위치 조정, 추가적으로 상수화 가능한 부분들 추가)
- [x] `getWinningCount` 함수에서 하나의 로또에 대한 로직을 분리했어요.
- [x] 꽝에 대한 테스트 케이스 추가
- [x] `getLottoNumber` 테스트 신뢰성 향상 -> 로또 넘버가 범위 내에서 가져오는지에 대한 테스트의 신뢰성을 향상했어요. (이 방식이 올바른지는 잘 모르겠어요...)
- [x] 조건문 분리 -> 카운트 증가, 게임 종료 함수 내부에 있는 조건문을 외부로 분리해서 카운트 증가, 게임 종료만 담당할 수 있게 수정했어요.
- [x] 로또 정렬 로직을 분리했어요.
- [x] `Calculator` 내부에 있던 당첨 결과 산정 로직을 `LottoCenter`라는 객체로 이동시켰어요.
- [x] `purchase` 내부에 있는 로직을 관련된 동작끼리 묶어 별도의 함수로 분리했어요.

## 추가로 궁금한 것들
- 파일 외부에서 사용되지 않는 함수를 테스트를 위해서 내보내기를 해야 할까요?
- 조건문을 처리하기 위한 별도의 `handler` 함수를 생성하는 것이 올바른 방법일까요? (증가 조건, 게임 종료 조건)
- 한 줄짜리 코드도 역할이 늘어난다면 분리하는 것이 좋은가
    - ex) 발행한 로또를 오름차순으로 정렬하는 로직

리뷰 주신 내용을 토대로 반영해 보았는데 제가 맞게 이해하고 반영한 건지 잘 모르겠어요... 🥲

woowa-euijinkk commented on src/constant/output.js:
```diff
@@ -1,6 +1,7 @@
 
-export default OUTPUT_MESSAGE;
\ No newline at end of file
+export default OUTPUT_MESSAGE;
```
EOL 제거 좋아요 👍 
페어는 EOL 설정이 필요하겠군요 ㅎㅎ

woowa-euijinkk commented on src/util/validate.js:
```diff
@@ -3,6 +3,8 @@ import PRICE from "../constant/price.js";
 import RESTART_ANSWER from "../constant/answer.js";
 import { LOTTO } from "../constant/lotto.js";
 
+const SPLITTER = ",";
```
validate 과 ',' 가 강한 결합이 생겼군요.
우려되는 지점은,
validate가 input 을 알고 있어야 한다는 거에요.
',' 이 들어옴을 알아야만 이 코드가 성립되는거죠. 그런 측면에서 강한 결합이라고 말씀을 드리고요!

그러면 단점이 뭘까요?~
input 에서 , 로 받지 않으면 모든 코드가 깨진다는 거에요. 그런 측면에서 ','는 input 이 가지고 있어야 합니다.
import 를 할지, 의존성 주입을 할지는 더 고민해보고 어떤 것이 좋아보이는지 공부해보고 알려줄 수 있나요?~ (의존성 주입이라는 말을 모른다면 간단하게 공부해보시길 추천드려요!)

ohgus replied:
저는 의존성 주입이 더 괜찮다고 생각합니다!

지금은 하나의 splitter를 사용하고 있지만 혹시 사용자가 splitter를 선택해서 사용하게 할 수도 있지 않을까요?
만약 그렇다면 그 상황에 맞게 splitter를 넘겨주는 것도 좋지 않을까 생각합니다!

하지만 만약 무조건 정해진 splitter로만 사용해야 한다면 import가 더 맞지 않을까 싶기도 한 것 같아요...🥲
그래도 저는 유연성을 위해서 의존성 주입 방식으로 수정하겠습니다 🤔

ohgus replied:
수정했습니다!

[SPLITTER 위치 변경](https://github.com/woowacourse/javascript-lotto/pull/363/commits/308153dfeb5a9f34545e802cd6ab2fe7f63ba96e)

woowa-euijinkk commented on __tests__/validate.test.js:
```diff
@@ -22,29 +23,29 @@ describe("validate", () => {
       test(`${LOTTO.MIN_RANDOM_VALUE}~${LOTTO.MAX_RANDOM_VALUE} 사이의 숫자가 아닌 경우`, () => {
         const numbers = [1, 2, 3, 4, 5, 6];
-        const number = "46";
+        const outOfRangeNumber = `${LOTTO.MAX_RANDOM_VALUE + 1}`;
```
👍

woowa-euijinkk commented on __tests__/validate.test.js:
```diff
@@ -22,29 +23,29 @@ describe("validate", () => {
       test(`중복된 번호가 있는 경우`, () => {
         const numbers = [1, 2, 3, 4, 5, 6];
-        const number = "6";
+        const duplicateNumber = "6";
```
```js
const duplicateNumber = "6";
const numbers = [1, 2, 3, 4, 5, duplicateNumber];
```

이런 식으로 하면 더 적절할 것 같습니다.

ohgus replied:
수정했습니다

[중복 변수화](https://github.com/woowacourse/javascript-lotto/pull/363/commits/4d303689ea591f99e919901996b6d6d8753e2e6f)

woowa-euijinkk commented on src/ui/OutputView.js:
```diff
@@ -25,7 +25,7 @@ const OutputView = {
 
   printYieldRate(yieldRate) {
-    console.log(`총 수익률은 ${yieldRate}%입니다.${NEW_LINE}`);
+    console.log(`총 수익률은 ${yieldRate.toFixed(1)}%입니다.${NEW_LINE}`);
```
👍

woowa-euijinkk commented on src/step1-index.js:
```diff
@@ -13,12 +13,18 @@ class App {
+
+  endGame() {
+    this.#running = false;
   }
```
while 문 때문에 2 depth 라 번거로워졌네요 ㅎㅎ
2 depth 로 짜는게 보편적일거같아요.

하드한 요구사항 하에서 잘 연습해보셨다고 생각해보세요.
이런 함수 분리 자체는 너무 훌륭합니다~!

woowa-euijinkk commented on src/LottoStore.js:
```diff
@@ -34,4 +27,25 @@ const readWinningInfo = async () => {
   };
 };
 
+const getLottos = (purchaseAmount) => {
+  const quantity = Calculator.getQuantity(purchaseAmount);
+  return Array.from({ length: quantity }, () => generateLotto());
+};
+
+const getYieldRate = (winningCounts, purchaseAmount) => {
+  const totalPrize = Calculator.getTotalPrize(winningCounts);
+
+  return Calculator.getYieldRate(purchaseAmount, totalPrize);
+};
+
+const showWinningDetail = (winningCounts) => {
+  OutputView.printWinningDetailTitle();
+  OutputView.printWinningDetail(winningCounts);
+};
+
+const showLottos = (lottos) => {
+  OutputView.printQuantity(lottos.length);
+  OutputView.printLottos(lottos);
+};
```
잘 분리해주신 것 같아요. 선언적이라 더 읽기가 쉽네요!

woowa-euijinkk reviewed (changes requested):
오거스~~ 
리뷰 많이 드렸는데, 잘 수정해주셨네요.

클린한 코드에 대해서 많이 고민해볼 수 있는 시간이 되셨다면 좋겠네요.
전체적인 구조도 구조지만, 일단 함수, 변수 하나하나를 신경 써서 짜보시는 연습은 계속 해보시면 좋습니다!
함수명, 인자명, 함수의 역할 등을 계속 고민해보는거죠!

이번에 훌륭하게 개선이 되어서 좋네요.

## 질문 1
> 파일 외부에서 사용되지 않는 함수를 테스트를 위해서 내보내기를 해야 할까요?

함수가 꼭 은닉이 필요한 기능이냐, 아니냐에 따라서 다를 것 같아요.
은닉할 필요가 없는 경우(중요한 데이터의 흐름을 바꾸지 않는 경우)라면 export 하든 말든 문제가 아니지 않을까요?
은닉이 필요한 경우는 어떻게 하면 좋을까요? 오거스의 의견이 궁금하네요

## 질문 2
> 조건문을 처리하기 위한 별도의 handler 함수를 생성하는 것이 올바른 방법일까요? (증가 조건, 게임 종료 조건)

조건문을 품는 함수는 저는 대체로 분리하지 않는 것 같긴 합니다?! 아래 형태로 자주 사용합니다. 
이번에 1depth 를 위해서, 많이 노력하셨을 텐데, 그 과정에서 얻는 것이 많았으리라 생각해요. 
현실적으로 언제나 1depth 를 유지할 수는 점! 

event handler 에는 당연히 분기가 포함될 수 있습니다. 브라우저가 정한 event 에 맞는 callback 을 날리는거니까요. 2단계 때 겪어보시죠!

```js
if (ok) {
  increase();
}
```

## 질문 3
> 한 줄짜리 코드도 역할이 늘어난다면 분리하는 것이 좋은가
ex) 발행한 로또를 오름차순으로 정렬하는 로직

저는 한줄도 바로 이해하기 어렵다면, 분리해야 한다고 생각합니다. 

아래 코드에서 두개 기능을 비교해보아요. 아래는 reduce 문을 끝까지 읽어야, "아 이건 다 더하는 함수구나", 라고 알 수 있는 것에 반해, 아래위는 그냥 앞만 봐도 sum 하는 함수구나 라고 이해할 수 있습니다. 

```js
const sum = arr.reduce((acc, curr) => acc + curr, 0)
const 총합 = sum([1,2,3])

// vs
const 총합 = [1,2,3].reduce((acc, curr) => acc + curr, 0)
```

말씀드린 sort 사례도 그렇다고 볼 수 있고요. 

---
너무 고생하셨어요! 거의 다 수정된 거 같아요.
오늘 내일중으로 빠르게 코멘트 주시면, 확인하고 바로 머지하고 2단계로 가시죠!

ohgus committed:
refactor: SPLITTER 위치 변경

- 별도 파일로 이동
- validate 함수에서 매개 변수로 splitter를 받을 수 있게 수정

ohgus committed:
refactor: 구입 금액을 숫자 타입으로 변환

ohgus committed:
refactor: 중복 체크 테스트 수정

ohgus committed:
rename: 파일명 오타 수정

ohgus committed:
refactor: 당첨 번호 중복 테스트 수정

ohgus committed:
refactor: 상금 계산을 하드 코딩하지 않게 수정

ohgus commented:
안녕하세요 시지프!
추가로 리뷰 주신 내용들 수정했습니다~
[여기부터 시작](https://github.com/woowacourse/javascript-lotto/pull/363/commits/308153dfeb5a9f34545e802cd6ab2fe7f63ba96e)

### 추가 질문 🙋
- 시지프가 개발자가 되고 싶었던 이유가 궁금해요 ㅎ
- 또 왜 프론트엔드를 선택하셨는지도 궁금합니다!
- 그리고 그 선택에 후회는 없으신가요?
- 시지프가 생각하는 프론트의 매력은 무엇이라고 생각하시나요?

시지프의 상세한 리뷰 덕분에 한 주 동안 많이 성장한 것 같아요 감사합니다!
2단계도 잘부탁드립니다 😁

woowa-euijinkk commented on __tests__/Calculator.test.js:
```diff
@@ -11,7 +12,8 @@ describe("Calculator test", () => {
     };
     const totalPrize = Calculator.getTotalPrize(winningCount);
-    const expectedTotalPrize = 3_000_000;
+    const expectedTotalPrize =
+      winningCount[KEYS.THIRD] * WINNING[KEYS.THIRD].PRIZES;
```
위에 0,2 숫자가 변경되면, 테스트가 바로 깨질 것 같습니다.
순회 후 더해야 합니다!

Object.entries 와 reduce 로 로직을 표현해보시면서, 객체, 배열 메서드 학습해보시죠.

ohgus replied:
제가 거기까지 생각하지 못한 것 같습니다 😅

이런 식으로 변경하면 괜찮을까요?
``` js
const expectedTotalPrize = Object.entries(winningCount).reduce(
      (acc, [key, value]) => acc + value * WINNING[key].PRIZES,
      0
    );
```

woowa-euijinkk replied:
잘 해주셨군요 👍

woowa-euijinkk reviewed (commented):
오거스 리뷰 반영하시느라 고생하셨어요.
1단계는 여기서 마무리 하겠습니다. 코멘트 2개 더 남겼는데, 한번 확인해보시고, 추가로 소통하고 싶으시면 해당 코멘트에 이어서 하셔도 됩니다.
2단계 하면서 고민해보셔도 되고요.

> 시지프가 개발자가 되고 싶었던 이유가 궁금해요 ㅎ
또 왜 프론트엔드를 선택하셨는지도 궁금합니다!
그리고 그 선택에 후회는 없으신가요?
시지프가 생각하는 프론트의 매력은 무엇이라고 생각하시나요?

개발자가 된 이유는, [이 블로그 글](https://happysisyphe.tistory.com/65) 에서 보실 수 있습니다 ㅎㅎ
저에게 개발은 제 목표를 이루기 위한 도구라서요, 프론트든 백엔드든 뭐든 크게 중요하진 않습니다 ㅎㅎ 
지금은 일단 둘다 어느정도 할 수 있고요. 
오거스도 몇년 하다보면, 이 기술이든 저 기술이든 크게 다르다고 느끼진 않을거에요. 

제가 생각하는 프론트만의 매력은, 소통을 많이 한다는 거같아요. 유저와도 간접적으로 소통하고, 팀 내에서도 서버, 디자인, 기획과 상대적으로 자주 소통하게 됩니다. 제품에 가까이서 기여할 수 있죠. 
그리고 프론트만으로도 빠르게 프로토타입이 가능한 제품이 많습니다. 스타트업에서는 프론트가 빛을 발할 수 있죠. 프론트가 누구냐에 따라서, 팀의 속도가 달라지는 것을 많이 느꼈어요. 

저는 UI 작업을 좋아하진 않아서, 그 부분은 단점이라고 생각합니다 ㅎㅎ 연차가 쌓인다면 그걸 하지 않긴 하겠죠? 

더 궁금한 점이 있다면, 커피챗이나 DM 으로 더 물어봐주세요 ㅎㅎ
2단계 때 또 봬요~
